### PR TITLE
Introduce memory limits for channels of packet batches

### PR DESCRIPTION
**Pull request by ckamm**
#### Problem

This continues #24961. Queues between packet processing stages used to not have memory limits.

#### Summary of Changes

- Adds a `BoundedPacketBatchChannel` and uses it to communicate packet batches between stages.
- Limits the number of packets that each stage receives from the channel in each iteration.

Thanks to @brennanwatt for a lot of cleanup.

The branch history is a mess (I had started cleaning it up but then Brennan started on top of the wip branch), let's squash it down.

@sakridge @jstarry @aeyakovenko 

EDIT: Updated message after fixes and improvements:

Introduce memory-limited channels for packet batches
 
 Previously all channels that connected
 - fetch stage
 - findpacketsenderstake stage
 - sigverify stage
 - banking stage
 were unbounded(). This created significant issues when more packets arrived than could be processed.
 
 Now, these channels have been replaced by BoundedPacketBatchChannel, which limits the number of packet batches that will be stored. When the channel overflows, newer inputs will be preferred over older batches.
 
 In tests, this patch allows a validator to run stably even when flooded with packets. Previously it would oscillate and eventually run out of memory. The tests sent roughly 3.8M packets/s, see details at
 https://github.com/solana-labs/solana/pull/24961#issuecomment-1119500335
 
 Furthermore, this change increases max throughput in bench-tps and doubles average throughput.
 before: Max=31k, AVG=10.3k, after: Max=36k, AVG=23k
 see https://github.com/solana-labs/solana/pull/25192#discussion_r872656907
 
 Lots of credit goes to @brennanwatt for discussion, cleanup, fixes, tests and review.

### CODE REVIEW

**Comment by ckamm** [comment]
This queue should get serious scrutiny: https://github.com/solana-labs/solana/pull/25192/files#diff-28653ad23bd2873faacb741fc3a95acb7e338f87e04d4459802efdce2b8d71dd

File: core/src/fetch_stage.rs [hunk 1]
```diff
@@ -85,40 +90,26 @@ impl FetchStage {
...
-        while let Ok(mut packet_batch) = recvr.try_recv() {
-            packet_batch.packets.iter_mut().for_each(mark_forwarded);
-            num_packets += packet_batch.packets.len();
-            packet_batches.push(packet_batch);
-            // Read at most 1K transactions in a loop
-            if num_packets > 1024 {
-                break;
-            }
-        }
+        let (mut packet_batches, num_packets) = recvr.recv(RECV_MAX_PACKETS)?;
```
Comment (ckamm) [inline core/src/fetch_stage.rs:101]: change to 1024 to preserve previous behavior

File: core/src/find_packet_sender_stake_stage.rs [hunk 2]
```diff
@@ -59,6 +61,7 @@ impl FindPacketSenderStakeStats {
                     self.apply_sender_stakes_time as i64,
                     i64
                 ),
+                ("max_out_queue", self.max_out_queue as i64, i64),
```
Comment (ckamm) [inline core/src/find_packet_sender_stake_stage.rs:64]: drop this, testing leftover

File: core/src/tpu.rs [hunk 3]
```diff
@@ -41,6 +44,12 @@ use {
 
 pub const DEFAULT_TPU_COALESCE_MS: u64 = 5;
 
+/// The default maximum number of batches in the queue that leaves the
+/// fetch stage.
+///
+/// 10k batches means up to 1.3M packets, roughly 1.6GB of memory
+pub const DEFAULT_TPU_MAX_QUEUED_BATCHES_UDP: usize = 10_000;
```
Comment (ckamm) [inline core/src/tpu.rs:51]: This value gets used for lots of channels now. Should we rename it to `DEFAULT_MAX_QUEUED_BATCHES_PER_CHANNEL` and apply it everywhere?

File: multinode-demo/common.sh [hunk 4]
```diff
@@ -53,11 +53,11 @@ else
       (
         set -x
         # shellcheck disable=SC2086 # Don't want to double quote
-        cargo $CARGO_TOOLCHAIN build $maybe_release --bin $program
+        ./cargo nightly $CARGO_TOOLCHAIN build $maybe_release --bin $program
```
Comment (ckamm) [inline multinode-demo/common.sh:56]: testing leftovers, remove!

File: core/src/tpu.rs (hunk 3 above)
Comment (brennanwatt) [reply core/src/tpu.rs:51]: Yes, agree. Renamed and moved into bounded_streamer.rs

File: core/src/fetch_stage.rs (hunk 1 above)
Comment (brennanwatt) [reply core/src/fetch_stage.rs:101]: Changed back to 1024

File: streamer/src/bounded_streamer.rs [hunk 5]
```diff
@@ -0,0 +1,449 @@
...
+    packet_count: usize,
+    max_queued_batches: usize,
+    sender_count: usize,
+}
+
+#[derive(Clone)]
+pub struct BoundedPacketBatchReceiver {
+    signal_receiver: Receiver<()>,
+    signal_sender: Sender<()>,
+    data: Arc<RwLock<PacketBatchChannelData>>,
```
Comment (sakridge) [inline streamer/src/bounded_streamer.rs:26]: oof, we need a lock here now?

Comment (brennanwatt) [reply streamer/src/bounded_streamer.rs:26]: Yeah.. every send/receive must acquire the lock for modifying the queue and batch/packet counts. Maybe we should explore using `bounded `channel and make sure all users handle cases here they get full.

Out of curiosity, I collected some bench tps pre/post these changes and numbers and the results were surprising. Max TPS looks about the same, but AVG TPS is significantly higher and more consistent with these changes.

Results testing each scenario 8x:
- Master: Max=31k, AVG=10.3k
- This branch: Max=31.6k, AVG=21.9k
*Note: this is the multinode benchmarking scripts w/ single bootstrap validator and client all running on the same GCE instance

Not clear to me if there's something else going on driving the AVG tps delta being observed here..

Testing iteration breakdown was performed serially like so:
1. 4x on this branch
2. 6x on master
3. 4x on this branch
4. 2x on master

Comment (ckamm) [reply streamer/src/bounded_streamer.rs:33]: I had originally looked at using `crossbeam_channel::bounded()` but:
- then you have to discard new data when its full, instead of discarding old
- the old code used a lot of `unbounded::<Vec<PacketBatch>>()` which are impossible to properly bound in terms of batch count

If we want to make the experiment, it should be easy to change `BoundedPacketBatchChannel` to be backed by a `bounded::<PacketBatch>()` while keeping the API the same. I don't expect much difference.

Comment (ckamm) [reply streamer/src/bounded_streamer.rs:33]: @brennanwatt I've made that hack/experiment in https://github.com/ckamm/solana/tree/ckamm/packetbatch-channel-limits-with-bounded - could you try running bench tps on that too?

Comment (brennanwatt) [reply streamer/src/bounded_streamer.rs:33]: I'm seeing solid results with the bounded chan experiment   - slightly better than with this branch:
- Max tps = 36k
- AVG tps = 23k

Comment (brennanwatt) [reply streamer/src/bounded_streamer.rs:33]: @sakridge what are we thinking as far as the path forward? The original code has really tight control at the expense of a lock. Bounded channel has looser control (although perhaps tight enough for practical use cases) but introduces no extra locking

Either way, having unified APIs for the different stages to control batches/packets count to receive is a great improvement

Comment (brennanwatt) [reply streamer/src/bounded_streamer.rs:33]: @sakridge what are we thinking as far as the path forward? The original code has really tight control at the expense of a lock. Bounded channel has looser control (although perhaps tight enough for practical use cases) but introduces no extra locking

Either way, having unified APIs for the different stages to control batches/packets count to receive is a great improvement

Comment (brennanwatt) [reply streamer/src/bounded_streamer.rs:33]: @sakridge what are we thinking as far as the path forward? The original code has really tight control at the expense of a lock. Bounded channel has looser control (although perhaps tight enough for practical use cases) but introduces no extra locking

Either way, having unified APIs for the different stages to control batches/packets count to receive is a great improvement

**Comment by brennanwatt** [comment]
@pgarg66 Making you aware of these changes as it impacts some of the banking stage tuning you've done

File: streamer/src/bounded_streamer.rs (hunk 5 above)
Comment (ckamm) [reply streamer/src/bounded_streamer.rs:33]: A third option is to rewrite in terms of a lock free queue, like `crossbeam::queue::ArrayQueue`. That should have similar performance while still allowing us to prefer new data.

**Comment by pgarg66** [comment]
> @pgarg66 Making you aware of these changes as it impacts some of the banking stage tuning you've done

Can we test the performance with QUIC? That was the reason for banking stage tuning, as the packet batching is/was different for QUIC.

Let me know if I can help with this testing..

**Comment by brennanwatt** [comment]
> > @pgarg66 Making you aware of these changes as it impacts some of the banking stage tuning you've done
> 
> Can we test the performance with QUIC? That was the reason for banking stage tuning, as the packet batching is/was different for QUIC.
> 
> Let me know if I can help with this testing..

I kicked off a bench-tps run with QUIC enabled. Let me know if there's anything specific you think we should be checking or if you can share your exact perf testing procedure

**Comment by pgarg66** [comment]
> > > @pgarg66 Making you aware of these changes as it impacts some of the banking stage tuning you've done
> > 
> > 
> > Can we test the performance with QUIC? That was the reason for banking stage tuning, as the packet batching is/was different for QUIC.
> > Let me know if I can help with this testing..
> 
> I kicked off a bench-tps run with QUIC enabled. Let me know if there's anything specific you think we should be checking or if you can share your exact perf testing procedure

That's a good test. We can also try to enable QUIC on the validator nodes. That'll make packet forwarding to use QUIC.

File: streamer/src/bounded_streamer.rs (hunk 5 above)
Comment (ckamm) [reply streamer/src/bounded_streamer.rs:33]: Here's a hacky implementation based on `crossbeam::queue::ArrayQueue`: https://github.com/ckamm/solana/tree/ckamm/packetbatch-channel-limits-with-crossbeam. @brennanwatt could you run bench-tps on that?

Comment (brennanwatt) [reply streamer/src/bounded_streamer.rs:33]: Performance looks good. In line with the bounded channel implementation

Comment (sakridge) [reply streamer/src/bounded_streamer.rs:33]: normal unbounded channel with an atomic to track the total packets in the channel could be an option. It will race, but we don't need to be exactly perfect.

Comment (ckamm) [reply streamer/src/bounded_streamer.rs]: Thanks for the test! I've cleaned it up and pushed it to this branch. Please review carefully.

**Comment by brennanwatt** [comment]
> > > > @pgarg66 Making you aware of these changes as it impacts some of the banking stage tuning you've done
> > > 
> > > 
> > > Can we test the performance with QUIC? That was the reason for banking stage tuning, as the packet batching is/was different for QUIC.
> > > Let me know if I can help with this testing..
> > 
> > 
> > I kicked off a bench-tps run with QUIC enabled. Let me know if there's anything specific you think we should be checking or if you can share your exact perf testing procedure
> 
> That's a good test. We can also try to enable QUIC on the validator nodes. That'll make packet forwarding to use QUIC.

Tested w/ QUIC enabled for both client and validator. I'm seeing similar performance on master vs. this branch:
- ~54k max TPS
- ~40k AVG TPS

File: streamer/src/bounded_streamer.rs [hunk 6]
```diff
@@ -0,0 +1,440 @@
...
+    // Returns (vec-of-batches, packet-count)
+    fn try_recv(
+        &self,
+        packet_count_hint: usize,
+    ) -> Result<(Vec<PacketBatch>, usize), TryRecvError> {
+        let mut packets = 0;
+        let disconnected = self.data.sender_count.load(Ordering::SeqCst) == 0;
+
+        let mut batches = Vec::with_capacity(min(
+            packet_count_hint / PACKETS_PER_BATCH,
```
Comment (brennanwatt) [inline streamer/src/bounded_streamer.rs:132]: Is this too optimistic? What about the case where we have very few packets in each batch?

File: core/src/banking_stage.rs [hunk 7]
```diff
@@ -2010,16 +1991,11 @@ impl BankingStage {
...
     ) -> Result<(), RecvTimeoutError> {
         let mut recv_time = Measure::start("receive_and_buffer_packets_recv");
-        let packet_batches = Self::receive_until(
-            verified_receiver,
-            recv_timeout,
-            Duration::from_millis(MAX_RECEIVE_TIME_MS_PER_ITERATION),
-            MAX_RECEIVE_BATCH_SIZE_PER_ITERATION,
-        )?;
+        let (packet_batches, packet_count) =
+            verified_receiver.recv_timeout(MAX_RECEIVE_PACKETS_PER_ITERATION, recv_timeout)?;
```
Comment (brennanwatt) [inline core/src/banking_stage.rs:1995]: @pgarg66 - Rather than receiving max of 50k batches, we will now receive max of 100k packets. I believe this change will allow for tighter arbitration (since counting batches is at the mercy of how dense the batches are), but it's not clear if 100k is the right number. Simple bench perf test seems to indicate we haven't completely screwed things up, but curious what your thoughts are on this

File: core/src/tpu.rs [hunk 8]
```diff
@@ -222,14 +255,16 @@ impl Tpu {
 
         Self {
             fetch_stage,
-            sigverify_stage,
-            vote_sigverify_stage,
+            udp_sigverify_stage,
```
Comment (sakridge) [inline core/src/tpu.rs:258]: can we do this renaming in a separate PR?

File: core/src/banking_stage.rs (hunk 7 above)
Comment (pgarg66) [reply core/src/banking_stage.rs:1995]: The newer code (tip of master and v1.10) is actually counting packets now. So the goals of this PR aligns well with what banking_stage is doing today.

File: core/src/tpu.rs (hunk 8 above)
Comment (brennanwatt) [reply core/src/tpu.rs:258]: Just to be clear, you're proposing we push the separate UDP vs. QUIC stage threads (but still keep unbounded channels as the packet passing mechanism) first as a separate PR, right?

Comment (ckamm) [reply core/src/tpu.rs:258]: I'll just move this commit to a separate PR? https://github.com/solana-labs/solana/pull/25192/commits/16a3b4a96e59979563880785d236e7e31c911d12

Comment (ckamm) [reply core/src/tpu.rs:258]: done as #25285

Comment (sakridge) [reply core/src/tpu.rs:258]: ah, sorry I didn't realize it's adding those new ones.. why do we need them?

Comment (ckamm) [reply core/src/tpu.rs:258]: We don't stricly need them. But when talking with @brennanwatt we realized that using one stage for QUIC and UDP would mean that a flood of UDP packets could crowd out QUIC packets. But QUIC packets are probably higher value. So we're now merging the flows only after sigverify.

If you like I can drop the new stages and instead merge in the queue leading to findpacketsenderstake.

File: client/tests/quic_client.rs [hunk 9]
```diff
@@ -24,7 +26,7 @@ mod tests {
         solana_logger::setup();
         let s = UdpSocket::bind("127.0.0.1:0").unwrap();
         let exit = Arc::new(AtomicBool::new(false));
-        let (sender, receiver) = unbounded();
+        let (sender, receiver) = packet_batch_channel(DEFAULT_MAX_QUEUED_BATCHES);
```
Comment (sakridge) [inline client/tests/quic_client.rs:29]: why do we need to change this?

File: core/src/tpu.rs (hunk 8 above)
Comment (brennanwatt) [reply core/src/tpu.rs]: Re-combined the UDP/QUIC TPU stages

File: core/src/tpu.rs [hunk 10]
```diff
@@ -169,7 +175,8 @@ impl Tpu {
             )
         };
 
-        let (verified_tpu_vote_packets_sender, verified_tpu_vote_packets_receiver) = unbounded();
+        let (verified_tpu_vote_packets_sender, verified_tpu_vote_packets_receiver) =
+            packet_batch_channel(50_000);
```
Comment (ckamm) [inline core/src/tpu.rs:179]: Right now there's a mix of
```
packet_batch_channel(tpu_max_queued_batches)
packet_batch_channel(DEFAULT_MAX_QUEUED_BATCHES)
packet_batch_channel(50_000)
```
in tpu.rs.

File: streamer/src/bounded_streamer.rs (hunk 6 above)
Comment (ckamm) [reply streamer/src/bounded_streamer.rs:132]: It's a new optimization: the old `recv_packet_batches()` didn't control the result vec capacity at all. So it should be an improvement even if this lower bound often underestimates the actual final capacity.

We could add a `* 2`. If 100k packets are requested, that'd shift the min from 781 to 1562 and probably cost around an extra 32kb of memory which may or may not be needed. I'm ambivalent and would probably keep as-as, but increasing this a bit is fine too.

If we wanted to go all-in on optimization, we'd receive into the same Vec every time to skip this reallocation in most cases, but I think the api-niceness cost is not worth the gain.

File: gossip/src/cluster_info.rs [hunk 11]
```diff
@@ -2822,8 +2815,12 @@ impl Node {
...
         let (tvu_port, tvu) = Self::bind(bind_ip_addr, port_range);
         let (tvu_forwards_port, tvu_forwards) = Self::bind(bind_ip_addr, port_range);
-        let ((tpu_port, tpu), (_tpu_quic_port, tpu_quic)) =
-            bind_two_consecutive_in_range(bind_ip_addr, port_range).unwrap();
+        let (tpu_port, tpu_sockets) =
+            multi_bind_in_range(bind_ip_addr, port_range, 32).expect("tpu multi_bind");
+        let (_tpu_port_quic, tpu_quic) = Self::bind(
+            bind_ip_addr,
+            (tpu_port + QUIC_PORT_OFFSET, tpu_port + QUIC_PORT_OFFSET + 1),
+        );
```
Comment (brennanwatt) [inline gossip/src/cluster_info.rs]: I've been seeing lots of bind failures in unit tests. Reverting this seems to have resolved things

Comment (ckamm) [reply gossip/src/cluster_info.rs]: yes, definitely revert that, this was not supposed to be in the PR at all

**Comment by ckamm** [comment]
@brennanwatt @sakridge I've squashed this down to one commit and think it's good to go. I still regularly see issues on mainnet that could plausibly be related to issues addressed by this PR. What is needed to make this go in?

**Review by jstarry — COMMENTED** [review]
Nits are optional, main concern is spreading work over multiple receivers because that's important in banking stage. Thanks for the contribution!

File: streamer/src/bounded_streamer.rs [hunk 12]
```diff
@@ -0,0 +1,440 @@
...
+        if let Some(b) = self.data.queue.pop() {
+            self.data.sub_packet_count(b.packets.len());
+            *discarded_batches += 1;
+        }
+    }
+
+    /// Pushes a batch, potentially discarding others to make it happen.
+    fn push(&self, mut batch: PacketBatch, discarded_batches: &mut usize) {
+        self.data.add_packet_count(batch.packets.len());
+        while let Err(uninserted_batch) = self.data.queue.push(batch) {
```
[the remark this answers is not in the document]
Comment (brennanwatt) [reply streamer/src/bounded_streamer.rs:265]: `force_push` makes sense to me

File: streamer/src/bounded_streamer.rs [hunk 13]
```diff
@@ -0,0 +1,440 @@
...
+/// Creates the sender and receiver for a channel of packet batches.
+///
+/// The channel is multi-producer multi-consumer.
+///
+/// It is bounded to hold at most `max_queued_batches`. If more batches are
+/// pushed into the channel, the oldest queued batches will be discarded.
+///
+/// It also counts the number of packets that are available in the channel,
+/// but is not aware of a Packet's meta.discard flag.
+pub fn packet_batch_channel(
```
[the remark this answers is not in the document]
Comment (ckamm) [reply streamer/src/bounded_streamer.rs:291]: maybe `streamer::packet_batch_channel::bounded`?

File: streamer/src/bounded_streamer.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (ckamm) [reply streamer/src/bounded_streamer.rs:265]: yeah, I looked at an outdated version of the docs and didn't see it :)

**Comment by behzadnouri** [comment]
Can we please reconsider if this is the right or the best approach?

* There are downstream processing stages that we know are much faster than earlier upstream stages. Channel size growth is not a problem there. Having a bounded channel in between the two only adds overhead.
* There are places which already do their own load-shedding but also track number of discarded packets through metrics; e.g.
https://github.com/solana-labs/solana/blob/021276c95/gossip/src/cluster_info.rs#L2454-L2463
https://github.com/solana-labs/solana/blob/021276c95/gossip/src/cluster_info.rs#L2498-L2507
moving the load shedding somewhere else will break those metrics, and limits our insight when debugging those processes.
* We also might want to later do a smarter load shedding than just FIFO or LIFO. For example gossip receives different packets and their processing computation cost are significantly different. Ideally when overloaded we can keep lighter or more important packets and discard packets which require heavy processing.
* Dropping packets just by FIFO or LIFO can open attack vectors. For example in turbine we can easily drop shreds which are duplicate or belong to an old slot. But if the packets are dropped before they even reach turbine, then it is easy to starve turbine of the right shreds by just sending too much invalid payload down the socket.
* `DEFAULT_MAX_QUEUED_BATCHES` is not a size fitting all; and the right optimal value depends on so many factors: specific downstream and upstream processing stages (e.g. gossip vs turbine), machine specs, networks bandwidth, node stake, etc ... I am worried this change will silently and unnecessary drop packets at so many places which will degrade cluster nodes and/or specific services such as gossip and turbine.
* We have had previous instances where silently dropping packets has been the source of very hard to debug bugs: https://github.com/solana-labs/solana/pull/16457


Overall, I am very against this approach and using bounded channels in general. I think the better alternative is that:
* Each stage sends the outgoing packets down an unbounded channel.
* The receiving end will each time drain the entire channel, and discards excess load however it deems reasonable for itself; similar to:
https://github.com/solana-labs/solana/blob/021276c95/gossip/src/cluster_info.rs#L2498-L2507
the receiving end has much more context for optimal load shedding strategy for itself. Potentially we can do something more optimal than FIFO or LIFO; and take into account the payload of those packets; and distinguish bad vs good packets.
* When dropping excess packets we can collect much more informative metrics which helps debugging. For example we can collect some stats about the characteristic of those packets when there is high spam rate or invalid payload.

**Comment by t-nelson** [comment]
where was the motivation for preferring new to old buffered entries elaborated?  seems like that would allow poorly behaved traffic to displace well behaved traffic unfairly

**Comment by sakridge** [comment]
Counter argument to that is, if the downstream is behaving well and clearing the channel fast enough then this logic should never get hit. We also can have a metric to report this behavior so it's not exactly silent. And dropping could be better than an ever-growing buffer which we then can look at the metrics to see how we can drain the channel and apply a faster or coarser load-shedding algorithm.

The channel limits could be configured per-service and potentially even for the machine.

**Comment by ckamm** [comment]
@behzadnouri Thanks, this is great feedback! (It would have been good to hear that three weeks ago when we were discussing approaches - but here we are. Appreciated nonetheless!)

- The key thing that really needs to happen is to introduce load-shedding before dedup, in sigverify. Currently its input channel grows huge and its iteration time spirals out of control. This is the source of bad performance on mainnet and in tests during packet floods.
- For safety, I'd want to do the same in fetchpacketsenderstake, though that seems to be fast enough usually.
- `banking_stage::receive_until` may also need to be looked at again: currently its channel could grow arbitrarily if it fills faster than the consumers can handle. And in its multi-consumer situation you don't want to greedy-receive until the channel is empty.

Generally, I think that unbounded channels are a liability. You either need backpressure or load shedding. For single consumers, an unbounded channel with consumer-side load shedding will work though.

**Comment by behzadnouri** [comment]
> The key thing that really needs to happen is to introduce load-shedding before dedup, in sigverify.

My suggestion is to lets start off by adding a load shedding or back pressure mechanism specifically there.
The problem with rolling out a global solution is that there is no solution which is globally suitable across all these protocols.

> Generally, I think that unbounded channels are a liability.

My point is that I am not sure this is the best approach to address that problem.



Can we instead start off by
1. adding load shedding right at the specific protocol entries which are known contention points; 
  i.e. drain the channel there, then drop whatever is excess load.
2. also add some metrics there tracking number of packets drained from the channel and number of packets discarded.

Then revisit this issue again and based on the metrics investigate if the load shedding there is doing the job or if we need
* smarter or more aggressive load shedding.
* or more load shedding at the earlier stages 
* or back pressure on the earlier stages.
* or some form of bounded channel some where between specific stages.

Though at some point it might make sense to use bounded channels at specific places in the pipeline, starting right away by adding bounded channels very early on and everywhere else [does not sound the best approach to me](https://github.com/solana-labs/solana/pull/25192#issuecomment-1132040969).

Also as [Trent mentioned](https://github.com/solana-labs/solana/pull/25192#issuecomment-1132169839) dropping old packets in favor of newer packets will encourage bad actors. But then if we don't have a preference for newer packets, then back pressure is a better alternative than load shedding because otherwise you take machine resources to do some processing on the packets only to drop them soon after, whereas back pressure will free up machine resources for the consumer to catch up with the producer.

So there are at least two open design questions which are both very protocol specific:
1. which packets we want to drop when overloaded; 
  in particular this should not encourage spammers, and ideally can do something smarter than just time based.
2. if we can instead put back pressure on the producers to free up system resources rather than wasting machine resources.

**Comment by brennanwatt** [comment]
We can start with more of a point solution and evolve from there.

Beyond the DoS concerns, I like the idea of having APIs for doing common things on the receiver side like “receive max of X packets and wait up to Y milliseconds” instead of reinventing these solutions in several places (tuning parameters can obviously still be stage specific). But this can be a separate PR

Regarding the design decisions, I see a few options for load shedding:
1. Drop the oldest (could reward spammers)
2. Drop the newest
3. Drop random
4. Balance across sender addresses
Any other ideas?

**Comment by behzadnouri** [comment]
> instead of reinventing these solutions in several places

The point was not copy pasting the same code at several places.
The point is that if a specific protocol (e.g. gossip) knows which packets are important, or easy to process, or outdated or if lifo is better than fifo, or if it should prioritize higher stake senders, etc ... then we can do much more suitable load-shedding/back-pressure there than at an earlier stage in the pipeline where **that extra necessary context simply does not exist**.

**Comment by brennanwatt** [comment]
> > instead of reinventing these solutions in several places
> 
> The point was not copy pasting the same code at several places. The point is that if a specific protocol (e.g. gossip) knows which packets are important, or easy to process, or outdated or if lifo is better than fifo, or if it should prioritize higher stake senders, etc ... then we can do much more suitable load-shedding/back-pressure there than at an earlier stage in the pipeline where **that extra necessary context simply does not exist**.

Yep, all that makes sense. I’m referring to something different.

Even with unbounded channels that don’t shed load, we often don’t like to drain the whole channel. Unifying these APIs ended up being a part of this PR even though it wasn’t the main intent. 

We can add some form of this back in with a different PR

**Comment by ckamm** [comment]
I think this has been effectively resolved by adding load shedding to the consumers.
